src/lib/tls/tls13/tls_cipher_state.cpp

```diff
@@ -114,12 +118,43 @@ std::unique_ptr<Cipher_State> Cipher_State::init_with_server_hello(
    const Ciphersuite& cipher,
    const Transcript_Hash& transcript_hash)
    {
-   auto cs = std::unique_ptr<Cipher_State>(new Cipher_State(side, cipher));
+   auto cs = std::unique_ptr<Cipher_State>(new Cipher_State(side, cipher.prf_algo()));
    cs->advance_without_psk();
-   cs->advance_with_server_hello(std::move(shared_secret), transcript_hash);
+   cs->advance_with_server_hello(cipher, std::move(shared_secret), transcript_hash);
    return cs;
    }
 
+std::unique_ptr<Cipher_State> Cipher_State::init_with_psk(
+         const Connection_Side side,
+         const Cipher_State::PSK_Type type,
+         secure_vector<uint8_t>&& psk,
+         const Ciphersuite& cipher)
+   {
+   auto cs = std::unique_ptr<Cipher_State>(new Cipher_State(side, cipher.prf_algo()));
```

Comment:
I read the RFC so that for PSK we just need to know the hash function from the start but during the remaining negotiation we might select a different cipher suite (that must share the same hash function, though). We should have tests that verify this possibility (maybe boGo has some?) and add a check in `advance_with_server_hello()` that the negotiated cipher suite doesn't change the pre-determined hash function.